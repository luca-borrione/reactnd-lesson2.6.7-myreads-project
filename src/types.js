
import PropTypes from 'prop-types'

/**
 * @typedef {Object} TBook
 * @description
 * Custom type to define a book internally.
 * It's filled with properties coming from the
 * book literal object fetched using the BooksAPI
 * @property {string[]} [authors] - List of the authors names
 * @property {string} id - The book unique id
 * @property {string} [shelf] - The shelf unique name
 * @property {string} [thumbnail] - URL for the cover preview image
 * @property {string} title - The book's title
 */

export const TBook = {
	authors: PropTypes.arrayOf(PropTypes.string.isRequired),
	id: PropTypes.string.isRequired,
	shelf: PropTypes.string,
	thumbnail: PropTypes.string,
	title: PropTypes.string.isRequired
};

