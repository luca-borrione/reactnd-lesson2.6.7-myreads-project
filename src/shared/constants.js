import { TShelfKey } from './types';

/**
 * @module
 * @name constants
 * @description
 * Provides a set of constants to be used in different components.
 */


/**
 * @member
 * @description
 * Dictionary of shelf labels
 * see [TShelfKey]{@link module:types.TShelfKey}
 * @type {Object.<TShelfKey, string>}
 * @property {TShelfKey} CURRENTLY_READING - Currently Reading
 * @property {TShelfKey} WANT_TO_READ - Want to Read
 * @property {TShelfKey} READ - Read
 * @property {TShelfKey} NONE - None
 * @constant
 */
export const SHELF_TITLE = Object.freeze({
	[TShelfKey.CURRENTLY_READING]: 'Currently Reading',
	[TShelfKey.WANT_TO_READ]: 'Want to Read',
	[TShelfKey.READ]: 'Read',
	[TShelfKey.NONE]: 'None',
});

export const ERROR = {
	BOOKS_API_FAILED: 'unexpected result from BooksAPI',
	BOOKS_IN_MULTI_SHELVES: 'the books passed are contained in multiple shelves',
	UNEXPECTED_STATUS: 'unexpected status'
};
