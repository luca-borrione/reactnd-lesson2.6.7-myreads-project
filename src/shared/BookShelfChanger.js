import React from 'react';
import PropTypes from 'prop-types';
import { SHELF_TITLE } from './constants';
import { TBook } from './types';


/**
 * @module
 * @name BookShelfChanger
 * @extends React.Component
 * @description
 * Shows a selection that allows the user to change the shelf of the book
 */
class BookShelfChanger extends React.Component {

	/**
	 * @member
	 * @name STATUS
	 * @description Collection of possible status
	 * @property {string} BUSY - The user selectd a different shelf for this book and the App is busy changing the shelf remotely
	 * @property {string} READY - Either the component received the current shelf in the props or it successfully changed it remotely
	 * @static
	 */
	static STATUS = {
		BUSY: 'STATUS.BUSY',
		READY: 'STATUS.READY'
	};
	/**
	 * @member
	 * @name propTypes
	 * @description Intended types of the props passed to the component
	 * @property {TBook} book - The book the changer is associated with
	 * @property {function} updateBookShelf - see [updateBookShelf]{@link module:App~updateBookShelf}
	 * @static
	 */
	static propTypes = {
		book: PropTypes.shape(TBook).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	constructor(props) {
		super(props);
		this.onSelectChange = this.onSelectChange.bind(this);
	}


	/**
	 * @member
	 * @name state
	 * @property {TShelfKey} selectedValue - the current selected shelf key
	 * @property {STATUS} status - one of the possible [STATUS]{@link module:BookShelfChanger.STATUS}
	 * @private
	 */
	state = {
		selectedValue: '',
		status: ''
	};


	// The shelf key coming from the props becomes stored internally as state
	static getDerivedStateFromProps(nextProps, prevState) {
		if (prevState.selectedValue === '') {
			return {
				selectedValue: nextProps.book.shelf,
				status: BookShelfChanger.STATUS.READY
			};
		}
		return null;
	}


	/**
	 * @method
	 * @description
	 * Handles the change event of the user selection.
	 * Firstly it updates the internal state, then it triggers the
	 * [updateBookShelf]{@link module:App~updateBookShelf} method of the App component
	 * to update the list of books in the shelves
	 * @param {SyntheticEvent} event
	 * @returns {void}
	 * @private
	 */
	onSelectChange(event) {
		const { STATUS } = this.constructor;
		const { book, updateBookShelf } = this.props;

		this.setState({
			selectedValue: event.target.value,
			status: STATUS.BUSY
		}, async () => {
			await updateBookShelf(book, this.state.selectedValue);

			// - Moving shelf in MySearch will cause the book to unmount,
			// in this case the new shelf will be passed throught the props.
			// - Moving shelf in the SearchPage won't cause the book to be umounted,
			// but the changer won't be renderer again either, so we need
			// to set change the status to trigger a re-render.
			// We need to check if the component is mounted to prevent triggering
			// an error in the console
			if (this._isMounted) {
				this.setState({
					status: STATUS.READY
				});
			}
		});
	}


	componentDidMount() {
		this._isMounted = true;
	}


	componentWillUnmount() {
		this._isMounted = false;
	}


	/**
	 * @method
	 * @name render
	 * @description
	 * Creates the selection containing all the available shelves.
	 * @returns {ReactElement}
	 * @private
	 */
	render() {
		const { STATUS } = this.constructor;
		const { status } = this.state;

		return (
			<div className="book-shelf-changer">
				{status === STATUS.BUSY && (
					<div id="loading"></div>
				)}
				<select value={this.state.selectedValue} onChange={this.onSelectChange}>
					<option value="move" disabled>Move to...</option>
					{Object.entries(SHELF_TITLE).map( ([ shelfKey, shelfTitle ], index) => (
						<option key={index} value={shelfKey}>{shelfTitle}</option>
					))}
				</select>
			</div>
		);
	}
}


export default BookShelfChanger;
