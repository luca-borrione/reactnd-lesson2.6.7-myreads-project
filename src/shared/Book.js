import React from 'react';
import BookShelfChanger from './BookShelfChanger';
import { makeTitle, ucwords } from '../utils/StringUtils';
import PropTypes from 'prop-types';
import { TBook } from './types';


/**
 * @module
 * @name Book
 * @extends React.Component
 * @description Displays the book's preview
 */
class Book extends React.Component {

	/**
	 * @member
	 * @name propTypes
	 * @description Intended types of the props passed to the component
	 * @property {TBook} book - The book to show
	 * @property {function} updateBookShelf - see [updateBookShelf]{@link module:App~updateBookShelf}
	 * @static
	 */
	static propTypes = {
		book: PropTypes.shape(TBook).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	/**
	 * @method
	 * @name render
	 * @description
	 * Shows a preview of the book and displays a selection to change shelf by rendering
	 * the [BookShelfChanger]{@link module:BookShelfChanger} component
	 * @returns {ReactElement}
	 * @private
	 */
	render() {
		const { book, updateBookShelf } = this.props;
		let thumbnail = 'https://books.google.co.uk/googlebooks/images/no_cover_thumb.gif';
		if (book.imageLinks && book.imageLinks.thumbnail) {
			thumbnail = book.imageLinks.thumbnail;
		}

		return (
			<li>
				<div className="book">
					<div className="book-top">
						<div className="book-cover" style={{
							width: 128,
							height: 193,
							backgroundSize: 'cover',
							backgroundImage: 'url('+ thumbnail +')'
						}}></div>

						<BookShelfChanger
							key={`${book.id}-changer`}
							book={book}
							updateBookShelf={updateBookShelf} />

					</div>
					<div className="book-title">{makeTitle(book.title)}</div>
					<div className="book-authors">
						{(book.authors || []).map( (author, index) => (
							<span className="author" key={index}>{ucwords(author)}</span>
						))}
					</div>
				</div>
			</li>
		);
	}
}


export default Book;
