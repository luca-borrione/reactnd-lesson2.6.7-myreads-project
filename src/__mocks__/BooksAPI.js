import * as data from './allBooks.json'
import { TShelfKey } from '../shared/types'

const DELAY = 10;

const getDelay = () => DELAY * (Math.floor(Math.random() * 10) + 1); // random between 1 - 10

export const MOCKED = true;

let allBooks;
export const __resetBooks = () => {
	allBooks = JSON.parse(JSON.stringify(data.books)); // cloning
};
__resetBooks();

const getBooksInShelves = () =>
	allBooks.filter( ({ shelf }) => shelf !== TShelfKey.NONE );


export const getAll = () => {
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(booksInShelves);
		}, delay);
	});
};

export const get = bookId => {
	const book = allBooks.find( ({ id }) => id === bookId );
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(book);
		}, delay);
	});
};

const formatUpdatedBooks = (books) => {
	const fBooks = {};

	Object.keys(TShelfKey).forEach( key => {
		const shelfKey = TShelfKey[key];
		if (shelfKey !== TShelfKey.NONE) {
			fBooks[shelfKey] = books
							.filter( ({ shelf }) => shelf === shelfKey )
							.map( book => book.id );
		}
	});

	return fBooks
};

export const update = (book, shelf) => {
	const existingBook = allBooks.find( ({ id }) => id === book.id );
	existingBook.shelf = shelf;
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(formatUpdatedBooks(booksInShelves));
		}, delay);
	});
};

export const search = (query, maxResults) => {
	const re = new RegExp(`^${query}`, 'i');
	const result = allBooks.filter( ({ title, authors }) => {
		return re.test(title);
	})
	.slice(0, maxResults);

	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(result);
		}, delay);
	});
};

