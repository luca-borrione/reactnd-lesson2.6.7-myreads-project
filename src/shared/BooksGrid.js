import React from 'react';
import Book from './Book';
import PropTypes from 'prop-types';
import { TBook } from './types';
import { diff as ArrayDiff} from '../utils/ArrayUtils';

/**
 * @module
 * @name BooksGrid
 * @extends React.Component
 * @description
 * Prints a list of books.
 * This module is used both in MyReads to show the books for each shelf,
 * and in the SearchPage to show the books retrieved by a search.
 */
class BooksGrid extends React.Component {

	/**
	 * @member
	 * @name propTypes
	 * @description Intended types of the props passed to the component
	 * @property {TBook[]} books - List of the books to show
	 * @property {function} updateBookShelf - see [updateBookShelf]{@link module:App~updateBookShelf}
	 * @static
	 */
	static propTypes = {
		books: PropTypes.arrayOf(
			PropTypes.shape(TBook).isRequired
		).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};

	/**
	 * @method
	 * @name shouldComponentUpdate
	 * @param {Object} nextProps - ignored as the module does not expect props
	 * @param {TBook[]} nextProps.books - foo
	 * @returns {boolean} - false if the books to show don't change
	 * @private
	 */
	shouldComponentUpdate(nextProps) {
		return ArrayDiff(this.props.books, nextProps.books).length > 0;
	}


	/**
	 * @method
	 * @name render
	 * @description
	 * Renders the [Book]{@link module:Book} component for each book in the given books list
	 * @returns {ReactElement}
	 * @private
	 */
	render() {
		const { books, updateBookShelf } = this.props;
		return (
			<ol className="books-grid">
				{books.map( (book, index) => (

					<Book key={index}
						book={book}
						updateBookShelf={updateBookShelf} />

				))}
			</ol>
		);
	}
}


export default BooksGrid;
