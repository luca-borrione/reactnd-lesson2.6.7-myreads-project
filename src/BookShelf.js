import React from 'react';
import BooksGrid from './BooksGrid';
import PropTypes from 'prop-types';
import { TBook } from './types';
import { ERROR, SHELF_TITLE } from './Constants';

/**
 * @class BookShelf
 * @extends React.Component
 * @classdesc
 * Prints the shelf title and shows the list of all the books in the shelf
 * @hideconstructor
 */
class BookShelf extends React.Component {

	/**
	 * @property {Object} propTypes - Intended types passed to the component
	 * @property {TBook[]} propTypes.books - List of all the books in the shelf
	 * @property {function} propTypes.updateBookShelf
	 * 	Method of the [App]{@link App#updateBookShelf} component for updating the list of books
	 * 	in the shelves
	 * @static
	 */
	static propTypes = {
		books: PropTypes.arrayOf(
			PropTypes.shape(TBook).isRequired
		).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	/**
	 * @description
	 * - Prints the shelf's title
	 * - Shows the list of books in the shelf by rendering the [BooksGrid]{@link BooksGrid} component
	 * @returns {ReactElement}
	 * @throws {Error} if the books passed to the component are contained in multiple shelves an exception will be thrown
	 */
	render() {
		const { books, updateBookShelf } = this.props;

		const shelfKeys = [
			...new Set(books.map( book => book.shelf ))
		];

		if (shelfKeys.length > 1) {
			throw new Error(ERROR.BOOKS_IN_MULTI_SHELVES, shelfKeys);
		}
		const shelfTitle = SHELF_TITLE[shelfKeys[0]];

		return (
			<div className="bookshelf">
				<h2 className="bookshelf-title">{shelfTitle}</h2>
				<div className="bookshelf-books">

					<BooksGrid
						books={books}
						updateBookShelf={updateBookShelf} />

				</div>
			</div>
		);
	}

}


export default BookShelf;
