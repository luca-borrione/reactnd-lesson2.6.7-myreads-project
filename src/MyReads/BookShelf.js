import React from 'react';
import BooksGrid from '../shared/BooksGrid';
import PropTypes from 'prop-types';
import { TBook } from '../shared/types';
import { ERROR, SHELF_TITLE } from '../shared/constants';

/**
 * @module
 * @name BookShelf
 * @extends React.Component
 * @description
 * Prints the shelf title and shows the list of all the books in the shelf
 */
class BookShelf extends React.Component {

	/**
	 * @member
	 * @name propTypes
	 * @description Intended types of the props passed to the component
	 * @property {TBook[]} books - List of all the books in the shelf
	 * @property {function} updateBookShelf - see [updateBookShelf]{@link module:App~updateBookShelf}
	 * @static
	 */
	static propTypes = {
		books: PropTypes.arrayOf(
			PropTypes.shape(TBook).isRequired
		).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	/**
	 * @method
	 * @name render
	 * @description
	 * - Prints the shelf's title
	 * - Shows the list of books in the shelf by rendering the [BooksGrid]{@link module:BooksGrid} component
	 * @returns {ReactElement}
	 * @throws {Error} if the books passed to the component are contained in multiple shelves an exception will be thrown
	 */
	render() {
		const { books, updateBookShelf } = this.props;

		const shelfKeys = [
			...new Set(books.map( book => book.shelf ))
		];

		if (shelfKeys.length > 1) {
			throw new Error(ERROR.BOOKS_IN_MULTI_SHELVES, shelfKeys);
		}
		const shelfTitle = SHELF_TITLE[shelfKeys[0]];

		return (
			<div className="bookshelf">
				<h2 className="bookshelf-title">{shelfTitle}</h2>
				<div className="bookshelf-books">

					<BooksGrid
						books={books}
						updateBookShelf={updateBookShelf} />

				</div>
			</div>
		);
	}

}


export default BookShelf;
