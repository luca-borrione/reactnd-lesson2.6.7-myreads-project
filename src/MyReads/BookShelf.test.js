import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import BookShelf from './BookShelf';
import * as BooksAPI from '../BooksAPI'; // mocked
import { TShelfKey } from '../shared/types';
import { ERROR } from '../shared/constants';

describe('BookShelf', () => {

	let book;
	let allBooks;

	const updateBookShelf = () => {};

	beforeEach( async () => {
		if (!book) {
			book = await BooksAPI.get('nggnmAEACAAJ');
		}
		if (!allBooks) {
			allBooks = await BooksAPI.getAll();
		}
	});


	it('renders without crashing', () => {
		const props = {
			books: [book],
			updateBookShelf
		};
		const div = document.createElement('div');
		ReactDOM.render(<BookShelf {...props} />, div);
		ReactDOM.unmountComponentAtNode(div);
	});


	it('renders correctly', () => {
		const props = {
			books: [book],
			updateBookShelf
		};
		const tree = TestRenderer
			.create(<BookShelf {...props} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});


	it('expects books containing the same shelf', async () => {
		const props = {
			books: allBooks.filter( ({ shelf }) => shelf === TShelfKey.WANT_TO_READ),
			updateBookShelf
		};

		const fn = () => shallow(<BookShelf {...props} />);
		expect(fn).not.toThrowError();
	});


	it('throws an error if it receives books containing different shelves', async () => {
		const props = {
			books: allBooks.filter( ({ shelf }) =>
				shelf === TShelfKey.WANT_TO_READ ||
				shelf === TShelfKey.CURRENTLY_READING),
			updateBookShelf
		};

		const fn = () => shallow(<BookShelf {...props} />);
		expect(fn).toThrowError(ERROR.BOOKS_IN_MULTI_SHELVES);
	});

	it('should contain BooksGrid as a child', () => {
		const props = {
			books: [book],
			updateBookShelf
		};
		const wrapper = mount(<BookShelf {...props} />);
		expect(wrapper.find('BooksGrid')).toHaveLength(1);
	});


	describe('props behaviour', () => {

		it('passes the prop updateBookShelf method as it is to the BooksGrid child', () => {
			const props = {
				books: [book],
				updateBookShelf: jest.fn()
			};
			const shelf = TShelfKey.WANT_TO_READ;

			const wrapper = mount(<BookShelf {...props} />);

			wrapper.find('BooksGrid').instance().props.updateBookShelf(book, shelf);
			expect(props.updateBookShelf).toHaveBeenCalledTimes(1);
			expect(props.updateBookShelf).toHaveBeenCalledWith(book, shelf);
		});

	});
});
