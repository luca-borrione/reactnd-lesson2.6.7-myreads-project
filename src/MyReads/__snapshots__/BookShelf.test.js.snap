// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookShelf renders correctly 1`] = `
<div
  className="bookshelf"
>
  <h2
    className="bookshelf-title"
  >
    Currently Reading
  </h2>
  <div
    className="bookshelf-books"
  >
    <ol
      className="books-grid"
    >
      <li>
        <div
          className="book"
        >
          <div
            className="book-top"
          >
            <div
              className="book-cover"
              style={
                Object {
                  "backgroundImage": "url(http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api)",
                  "backgroundSize": "cover",
                  "height": 193,
                  "width": 128,
                }
              }
            />
            <div
              className="book-shelf-changer"
            >
              <select
                onChange={[Function]}
                value="currentlyReading"
              >
                <option
                  disabled={true}
                  value="move"
                >
                  Move to...
                </option>
                <option
                  value="currentlyReading"
                >
                  Currently Reading
                </option>
                <option
                  value="wantToRead"
                >
                  Want to Read
                </option>
                <option
                  value="read"
                >
                  Read
                </option>
                <option
                  value="none"
                >
                  None
                </option>
              </select>
            </div>
          </div>
          <div
            className="book-title"
          >
            The Linux Command Line
          </div>
          <div
            className="book-authors"
          >
            <span
              className="author"
            >
              William E. Shotts, Jr.
            </span>
          </div>
        </div>
      </li>
    </ol>
  </div>
</div>
`;
